augur/clades.py

```diff
@@ -147,16 +144,40 @@ def assign_clades(clade_designations, all_muts, tree, ref=None):
         sorted_nodes = sorted(node_counts, key=lambda x: x.leaf_count, reverse=True)
         if len(sorted_nodes) > 0:
             target_node = sorted_nodes[0]
-            clade_membership[target_node.name] = {'clade_annotation': clade_name, 'clade_membership': clade_name}
+            basal_clade_nodes[target_node.name] = clade_name
+            clade_membership[target_node.name] = clade_name # basal nodes are members of the clade
 
     # third pass to propagate 'clade_membership'
     # don't propagate if encountering 'clade_annotation'
     for node in tree.find_clades(order = 'preorder'):
         for child in node:
-            if 'clade_annotation' not in clade_membership[child.name]:
-                clade_membership[child.name]['clade_membership'] = clade_membership[node.name]['clade_membership']
+            # if the child doesn't define the start of its own clade, but the parent belongs to a clade, then inherit that membership
+            if child.name not in basal_clade_nodes and node.name in clade_membership:
+                clade_membership[child.name] = clade_membership[node.name]
+    return (basal_clade_nodes, clade_membership)
+
 
-    return clade_membership
+def create_node_data_structure(basal_clade_nodes, clade_membership, args):
+    node_data = {}
+
+    if (not args.label_name and not args.trait_name):
+        print("WARNING: running `augur clades` without specifying --label-name and/or")
+        print("         --trait-name is deprecated. To preserve backwards compatibility")
+        print("         we will use 'clade' and 'clade_membership', respectively.")
+        print("         (Note that 'clade' is now exported as a 'branch_label')")
+
+        label_name = "clade"
+        trait_name = "clade_membership"
+    else:
+        label_name = args.label_name
+        trait_name = args.trait_name
+
+    if trait_name:
+        node_data['nodes'] = {node: {trait_name: clade} for node,clade in clade_membership.items()}
+    if label_name:
+        node_data['branch_labels'] = {node: {label_name: clade} for node,clade in basal_clade_nodes.items()}
```

Comment:
It's really cool that we can build branch labels like this! After I realized that I could build "node data" for anything I wanted, I started doing it all the time in my own analyses. Is there a place we already document this node data functionality that we could also document the "branch labels" functionality?

Reply:
Not that i'm aware of. There's no schema, and our `NodeDataFile` class (which actually reads the file) will accept any keys (there's a method which will remove the `generated_by` key if called).

As fas as I can tell these top level properties are used rather ad-hoc, a quick scan of some nCoV intermediates revealed `mask`, `params`, `annotations`, `generated_by`, `reference`, `models`, `alignment`, `clock`, `input_tree` and `skyline`.

Reply:
(will be) resolved by https://github.com/nextstrain/docs.nextstrain.org/issues/58